novawallet/Modules/DApp/DAppBrowser/Tabs/DAppBrowserTabManager/DAppBrowserTabManager.swift

```diff
@@ -0,0 +1,369 @@
+}
+
+// MARK: Private
+
+private extension DAppBrowserTabManager {
+    func setup() {
+        browserTabProvider = subscribeToBrowserTabs(nil)
+    }
+
+    func clearInMemory() {
+        dAppTransportStates = [:]
+        tabs = [:]
+    }
+
+    func saveRenderWrapper(
+        renderData: Data?,
+        tabId: UUID
+    ) -> CompoundOperationWrapper<Void> {
+        if let renderData {
+            fileRepository.saveRenderOperation(for: tabId, data: { renderData })
+        } else {
+            fileRepository.removeRender(for: tabId)
+        }
+    }
+
+    func saveWrapper(for tab: DAppBrowserTab) -> CompoundOperationWrapper<DAppBrowserTab> {
+        let persistenceModel = tab.persistenceModel
+
+        let saveTabOperation = repository.saveOperation(
+            { [persistenceModel] },
+            { [] }
+        )
+
+        let resultOperation = ClosureOperation { [weak self] in
+            _ = try saveTabOperation.extractNoCancellableResultData()
+
+            self?.dAppTransportStates[tab.uuid] = tab.transportStates
+
+            return tab
+        }
+        resultOperation.addDependency(saveTabOperation)
+
+        return CompoundOperationWrapper(
+            targetOperation: resultOperation,
+            dependencies: [saveTabOperation]
+        )
+    }
+
+    func retrieveWrapper(for tabId: UUID) -> CompoundOperationWrapper<DAppBrowserTab?> {
+        if let currentTab = tabs[tabId] {
+            return .createWithResult(currentTab)
+        } else {
+            let fetchTabOperation = repository.fetchOperation(
+                by: { tabId.uuidString },
+                options: RepositoryFetchOptions()
+            )
+
+            let resultOperation = ClosureOperation<DAppBrowserTab?> { [weak self] in
+                guard
+                    let self,
+                    let fetchResult = try fetchTabOperation.extractNoCancellableResultData()
+                else {
+                    return nil
+                }
+
+                return map(persistenceModel: fetchResult)
+            }
+
+            resultOperation.addDependency(fetchTabOperation)
+
+            return CompoundOperationWrapper(
+                targetOperation: resultOperation,
+                dependencies: [fetchTabOperation]
+            )
+        }
+    }
+
+    func updateRenderWrapper(
+        render: DAppBrowserTabRenderProtocol,
+        tab: DAppBrowserTab
+    ) -> CompoundOperationWrapper<Void> {
+        let renderDataOperation = render.serializationOperation()
+
+        let wrapper: CompoundOperationWrapper<Void> = OperationCombiningService.compoundNonOptionalWrapper(
+            operationManager: OperationManager(operationQueue: operationQueue)
+        ) { [weak self] in
+            guard let self else {
+                throw DAppBrowserTabManagerError.renderCacheFailed
+            }
+
+            let renderData = try renderDataOperation.extractNoCancellableResultData()
+
+            return saveRenderWrapper(
```

Comment:
I would make renderData parameter - closure and get rid of wrapping with combining service since save file operation also accept closure

Reply:
The `renderDataOperation` function returns an optional `Data` value. In this case, we should clear the persisted render. The `saveRenderWrapper` method decides whether to return a save wrapper or clear wrapper depending on the `renderDataOperation` result. We cannot simply pass the closure further.